warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/job/reindex/ShardReindexMapper.java

```diff
@@ -0,0 +1,764 @@
+            if (this.batchMode != BatchMode.NONE) {
+                batchValues = new HashMap<>();
+            }
+        }
+    }
+
+    @Override
+    protected void cleanup(Context context) throws IOException, InterruptedException {
+        // process any remaining batch data
+        if (this.batchMode != BatchMode.NONE && this.batchValues.size() > 0) {
+            try {
+                processBatch(context);
+            } catch (IOException | InterruptedException e) {
+                throw new RuntimeException("Could not process final batch", e);
+            }
+        }
+
+        if (this.generateMetadata && this.indexHandler.getMetadata() != null) {
+            for (BulkIngestKey bik : this.indexHandler.getMetadata().getBulkMetadata().keySet()) {
+                for (Value v : this.indexHandler.getMetadata().getBulkMetadata().get(bik)) {
+                    context.write(bik, v);
+                }
+            }
+
+        }
+
+        // output counters if used
+        if (this.enableReindexCounters) {
+            for (String counterGroup : this.counters.keySet()) {
+                Map<String,Long> groupCounters = this.counters.get(counterGroup);
+                for (String counter : groupCounters.keySet()) {
+                    Long value = groupCounters.get(counter);
+                    // optionally dump the counters to logs
+                    if (this.dumpCounters) {
+                        log.info("COUNTER " + counterGroup + " " + counter + " " + value);
+                    } else {
+                        context.getCounter(counterGroup, counter).increment(value);
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    protected void map(Key key, Value value, Context context) throws IOException, InterruptedException {
+        // TODO rewrite all of this to make sense
+        // This is all a bit awkward since DatawaveKey already parses the Key, but is in the datawave-query-core package. Minimally parse the Key for three
+        // purposes
+        // 1. Is it an FI key. Ranges should be created to defeat non fi ranges, but no verify and skip when possible
+        // 2. Get the field name. Necessary for checking how a field is indexed
+        // 3. Get the data type. The data type is used to get the correct IngestHelperInterface which is used to make decisions regarding how a field is
+        // indexed. This may vary from data type to data type
+
+        // this is required for the partitioner, see EventMapper
+        MultiTableRangePartitioner.setContext(context);
+
+        // ensure the key is an fi
+        final byte[] cf = key.getColumnFamilyData().getBackingArray();
+        final int fiBaseLength = FI_START_BYTES.length;
+        if (cf.length <= fiBaseLength || WritableComparator.compareBytes(cf, 0, fiBaseLength, FI_START_BYTES, 0, fiBaseLength) != 0) {
```

Comment:
This conditional tree is sorta tricky. It reads "if this isn't an FI, then check for tf, then check for d, if its not a tf or a d, it must be an event, else its an fi"

Wondering if we could bubble all the key checks to a single level and then maybe follow your `processFI` pattern for each of the other key types:

```
if (isFiCol(key))
  processFiCol(key, context)
else if (isTfCol(key))
  processTfCol(key, context)
else if (isDCol(key))
  processDCol(key, context)
else
  processEventKey(key, context)
```

Reply:
I will do this, but will need to add some new state to prevent reparsing